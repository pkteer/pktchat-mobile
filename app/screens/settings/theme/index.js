// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

import {savePreferences} from '@mm-redux/actions/preferences';
import {getTheme} from '@mm-redux/selectors/entities/preferences';
import {getCurrentTeamId} from '@mm-redux/selectors/entities/teams';
import {getCurrentUserId} from '@mm-redux/selectors/entities/users';
import {isLandscape, isTablet} from '@selectors/device';
import {getAllowedThemes, getCustomTheme} from '@selectors/theme';

import Theme from './theme';

const mapStateToProps = (state) => ({
    allowedThemes: getAllowedThemes(state),
    customTheme: getCustomTheme(state),
    isLandscape: isLandscape(state),
    isTablet: isTablet(state),
    teamId: getCurrentTeamId(state),
    theme: getTheme(state),
    userId: getCurrentUserId(state),
});

const mapDispatchToProps = (dispatch) => ({
    actions: bindActionCreators({
        savePreferences,
    }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Theme);
