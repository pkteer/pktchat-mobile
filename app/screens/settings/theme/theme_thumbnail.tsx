// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import Svg, {Rect, G, Circle} from 'react-native-svg';

type ThemeThumbnailProps = {
    width: number;
    borderColorBase: string;
    borderColorMix: string;
    sidebarBg: string;
    sidebarText: string;
    sidebarUnreadText: string;
    onlineIndicator: string;
    awayIndicator: string;
    dndIndicator: string;
    centerChannelColor: string;
    centerChannelBg: string;
    newMessageSeparator: string;
    buttonBg: string;
}

function ThemeThumbnail({
    width = 180,
    borderColorBase = '#E0E1E3',
    borderColorMix = '#E0E1E3',
    sidebarBg = '#174AB5',
    sidebarText = '#86A1D9',
    sidebarUnreadText = 'white',
    onlineIndicator = '#3DB887',
    awayIndicator = '#FFBC1F',
    dndIndicator = '#D24B4E',
    centerChannelColor = '#E0E1E3',
    centerChannelBg = 'white',
    newMessageSeparator = '#1C58D9',
    buttonBg = '#15B7B7',
}: ThemeThumbnailProps): JSX.Element {
    // the original height of the thumbnail
    const baseWidth = 180;
    const baseHeight = 134;

    // calculate actual height proportionally to base size
    const height = Math.round((width * baseHeight) / baseWidth);

    // convenience values of various sub elements of the thumbnail
    const sidebarWidth = 80;
    const postsContainerWidth = 100;
    const spacing = 8;
    const rowHeight = 6;
    const rowRadius = rowHeight / 2;
    const postInputHeight = 10;
    const postWidth = postsContainerWidth - (spacing * 2);
    const channelNameWidth = sidebarWidth - (spacing * 3) - (rowHeight * 2);
    const buttonWidth = postsContainerWidth - (spacing * 8);

    return (
        <Svg
            width={width}
            height={height}
            viewBox={`-2 -2 ${baseWidth + 4} ${baseHeight + 4}`}
            fill='none'
        >
            <Rect
                fill={centerChannelBg}
                x='0'
                y='0'
                width={baseWidth}
                height={baseHeight}
            />
            <Rect
                fill={newMessageSeparator}
                x={sidebarWidth}
                y={(spacing * 4) + (rowHeight * 3)}
                width={postsContainerWidth}
                height='1'
            />
            <Rect
                fill={buttonBg}
                x={sidebarWidth + (spacing * 4)}
                y={(spacing * 8) + (rowHeight * 6) + 1}
                width={buttonWidth}
                height={rowHeight}
                rx={rowRadius}
            />
            <Rect
                fill={centerChannelColor}
                x={sidebarWidth + spacing}
                y={(spacing * 9) + (rowHeight * 7) + 1}
                width={postWidth}
                height={postInputHeight}
                rx={postInputHeight / 2}
            />
            <Rect
                fill={centerChannelBg}
                x={sidebarWidth + spacing + 1}
                y={(spacing * 9) + (rowHeight * 7) + 2}
                width={postWidth - 2}
                height={postInputHeight - 2}
                rx={(postInputHeight - 2) / 2}
            />
            <G fill={centerChannelColor}>
                <Rect
                    x={sidebarWidth + spacing}
                    y={spacing}
                    width={postWidth}
                    height={rowHeight}
                    rx={rowRadius}
                />
                <Rect
                    x={sidebarWidth + spacing}
                    y={(spacing * 2) + rowHeight}
                    width={postWidth}
                    height={rowHeight}
                    rx={rowRadius}
                />
                <Rect
                    x={sidebarWidth + spacing}
                    y={(spacing * 3) + (rowHeight * 2)}
                    width={postWidth}
                    height={rowHeight}
                    rx={rowRadius}
                />
                <Rect
                    x={sidebarWidth + spacing}
                    y={(spacing * 5) + (rowHeight * 3) + 1}
                    width={postWidth}
                    height={rowHeight}
                    rx={rowRadius}
                />
                <Rect
                    x={sidebarWidth + spacing}
                    y={(spacing * 6) + (rowHeight * 4) + 1}
                    width={postWidth}
                    height={rowHeight}
                    rx={rowRadius}
                />
                <Rect
                    x={sidebarWidth + spacing}
                    y={(spacing * 7) + (rowHeight * 5) + 1}
                    width={postWidth}
                    height={rowHeight}
                    rx={rowRadius}
                />
            </G>
            <G>
                <Rect
                    fill={sidebarBg}
                    x='0'
                    y='0'
                    width={sidebarWidth}
                    height={baseHeight}
                />
                <G fill={sidebarText}>
                    <Circle
                        cx={spacing + rowRadius}
                        cy={spacing + rowRadius}
                        r={rowRadius}
                    />
                    <Circle
                        cx={spacing + rowRadius}
                        cy={(spacing * 2) + rowHeight + rowRadius}
                        r={rowRadius}
                    />
                    <Circle
                        cx={spacing + rowRadius}
                        cy={(spacing * 4) + (rowHeight * 3) + rowRadius}
                        r={rowRadius}
                    />
                    <Circle
                        cx={spacing + rowRadius}
                        cy={(spacing * 5) + (rowHeight * 4) + rowRadius}
                        r={rowRadius}
                    />
                    <Circle
                        cx={spacing + rowRadius}
                        cy={(spacing * 7) + (rowHeight * 6) + rowRadius}
                        r={rowRadius}
                    />
                    <Circle
                        cx={spacing + rowRadius}
                        cy={(spacing * 8) + (rowHeight * 7) + rowRadius}
                        r={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={spacing}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 2) + rowHeight}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 4) + (rowHeight * 3)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 5) + (rowHeight * 4)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 6) + (rowHeight * 5)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 7) + (rowHeight * 6)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 8) + (rowHeight * 7)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 9) + (rowHeight * 8)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                </G>
                <Circle
                    fill={onlineIndicator}
                    cx={spacing + rowRadius}
                    cy={(spacing * 3) + (rowHeight * 2) + rowRadius}
                    r={rowRadius}
                />
                <Circle
                    fill={awayIndicator}
                    cx={spacing + rowRadius}
                    cy={(spacing * 6) + (rowHeight * 5) + rowRadius}
                    r={rowRadius}
                />
                <Circle
                    fill={dndIndicator}
                    cx={spacing + rowRadius}
                    cy={(spacing * 9) + (rowHeight * 8) + rowRadius}
                    r={rowRadius}
                />
                <G fill={sidebarUnreadText}>
                    <Circle
                        cx={(spacing * 2.5) + rowHeight + channelNameWidth}
                        cy={(spacing * 3) + (rowHeight * 2) + rowRadius}
                        r={rowRadius}
                    />
                    <Rect
                        x={(spacing * 1.5) + rowHeight}
                        y={(spacing * 3) + (rowHeight * 2)}
                        width={channelNameWidth}
                        height={rowHeight}
                        rx={rowRadius}
                    />
                </G>
            </G>
            <Rect
                x='-1'
                y='-1'
                width={baseWidth + 2}
                height={baseHeight + 2}
                rx='4'
                stroke={borderColorBase}
                strokeWidth='2'
            />
            <Rect
                x='-1'
                y='-1'
                width={baseWidth + 2}
                height={baseHeight + 2}
                rx='4'
                stroke={borderColorMix}
                strokeWidth='2'
            />
        </Svg>
    );
}

export default ThemeThumbnail;
