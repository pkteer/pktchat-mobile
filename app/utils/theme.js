// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {StyleSheet} from 'react-native';
import tinyColor from 'tinycolor2';

import {mergeNavigationOptions} from '@actions/navigation';
import * as ThemeUtils from '@mm-redux/utils/theme_utils';

const MODAL_SCREENS_WITHOUT_BACK = [
    'AddReaction',
    'ChannelInfo',
    'CreateChannel',
    'EditPost',
    'ErrorTeamsList',
    'MoreChannels',
    'MoreDirectMessages',
    'Permalink',
    'SelectTeam',
    'Settings',
    'TermsOfService',
    'UserProfile',
];

export function makeStyleSheetFromTheme(getStyleFromTheme) {
    return ThemeUtils.makeStyleFromTheme((theme) => {
        return StyleSheet.create(getStyleFromTheme(theme));
    });
}

export const changeOpacity = ThemeUtils.changeOpacity;

export const blendColors = ThemeUtils.blendColors;

export function concatStyles(...styles) {
    return [].concat(styles);
}

export function setNavigatorStyles(componentId, theme) {
    const options = {
        topBar: {
            title: {
                color: theme.sidebarHeaderTextColor,
            },
            background: {
                color: theme.sidebarHeaderBg,
            },
            leftButtonColor: theme.sidebarHeaderTextColor,
            rightButtonColor: theme.sidebarHeaderTextColor,
        },
        layout: {
            componentBackgroundColor: theme.centerChannelBg,
        },
    };

    if (!MODAL_SCREENS_WITHOUT_BACK.includes(componentId)) {
        options.topBar.backButton = {
            color: theme.sidebarHeaderTextColor,
        };
    }

    mergeNavigationOptions(componentId, options);
}

export function isThemeSwitchingEnabled(state) {
    const {config} = state.entities.general;
    return config.EnableThemeSelection === 'true';
}

export function getKeyboardAppearanceFromTheme(theme) {
    return tinyColor(theme.centerChannelBg).isLight() ? 'light' : 'dark';
}

export function hexToHue(hexColor) {
    let {red, green, blue} = ThemeUtils.getComponents(hexColor);
    red /= 255;
    green /= 255;
    blue /= 255;

    const channelMax = Math.max(red, green, blue);
    const channelMin = Math.min(red, green, blue);
    const delta = channelMax - channelMin;
    let hue = 0;

    if (delta === 0) {
        hue = 0;
    } else if (channelMax === red) {
        hue = ((green - blue) / delta) % 6;
    } else if (channelMax === green) {
        hue = ((blue - red) / delta) + 2;
    } else {
        hue = ((red - green) / delta) + 4;
    }

    hue = Math.round(hue * 60);

    if (hue < 0) {
        hue += 360;
    }

    return hue;
}
