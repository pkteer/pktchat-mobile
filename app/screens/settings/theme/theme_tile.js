// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import PropTypes from 'prop-types';
import React from 'react';
import {
    Dimensions,
    TouchableOpacity,
    View,
} from 'react-native';

import CompassIcon from '@components/compass_icon';
import {changeOpacity} from '@mm-redux/utils/theme_utils';
import {makeStyleSheetFromTheme} from '@utils/theme';

import ThemeThumbnail from './theme_thumbnail';

const tilePadding = 8;

const ThemeTile = (props) => {
    const {
        action,
        actionValue,
        isLandscape,
        isTablet,
        label,
        selected,
        activeTheme,
        tileTheme,
    } = props;

    const style = getStyleSheet(activeTheme);

    const labelComponent = React.cloneElement(
        label,
        {style: style.label},
    );

    const tilesPerLine = isLandscape || isTablet ? 4 : 2;
    const {width: deviceWidth} = Dimensions.get('window');
    const fullWidth = isLandscape ? deviceWidth - 40 : deviceWidth;
    const layoutStyle = {
        container: {
            width: (fullWidth / tilesPerLine) - tilePadding,
        },
        thumbnail: {
            width: (fullWidth / tilesPerLine) - (tilePadding + 16),
        },
    };

    return (
        <TouchableOpacity
            style={[style.container, layoutStyle.container]}
            onPress={() => action(actionValue)}
        >
            <View style={[style.imageWrapper, layoutStyle.thumbnail]}>
                <ThemeThumbnail
                    width={layoutStyle.thumbnail.width}
                    borderColorBase={selected ? activeTheme.sidebarTextActiveBorder : activeTheme.centerChannelBg}
                    borderColorMix={selected ? activeTheme.sidebarTextActiveBorder : changeOpacity(activeTheme.centerChannelColor, 0.16)}
                    sidebarBg={tileTheme.sidebarBg}
                    sidebarText={changeOpacity(tileTheme.sidebarText, 0.48)}
                    sidebarUnreadText={tileTheme.sidebarUnreadText}
                    sidebarTextActiveBorder={activeTheme.sidebarTextActiveBorder}
                    onlineIndicator={tileTheme.onlineIndicator}
                    awayIndicator={tileTheme.awayIndicator}
                    dndIndicator={tileTheme.dndIndicator}
                    centerChannelColor={changeOpacity(tileTheme.centerChannelColor, 0.16)}
                    centerChannelBg={tileTheme.centerChannelBg}
                    newMessageSeparator={tileTheme.newMessageSeparator}
                    buttonBg={tileTheme.buttonBg}
                />
                {selected && (
                    <CompassIcon
                        name='check-circle'
                        size={31.2}
                        style={style.check}
                    />
                )}
            </View>
            {labelComponent}
        </TouchableOpacity>
    );
};

ThemeTile.propTypes = {
    action: PropTypes.func,
    actionValue: PropTypes.string,
    isLandscape: PropTypes.bool.isRequired,
    isTablet: PropTypes.bool.isRequired,
    label: PropTypes.node.isRequired,
    selected: PropTypes.bool,
    activeTheme: PropTypes.object.isRequired,
    tileTheme: PropTypes.object.isRequired,
};

ThemeTile.defaultProps = {
    action: () => true,
};

const getStyleSheet = makeStyleSheetFromTheme((theme) => {
    return {
        container: {
            flexDirection: 'column',
            padding: tilePadding,
            marginTop: 8,
        },
        imageWrapper: {
            position: 'relative',
            alignItems: 'flex-start',
            marginBottom: 12,
        },
        thumbnail: {
            resizeMode: 'stretch',
        },
        check: {
            position: 'absolute',
            right: 5,
            bottom: 5,
            color: theme.sidebarTextActiveBorder,
        },
        label: {
            color: theme.centerChannelColor,
            fontSize: 15,
        },
    };
});

export default ThemeTile;
