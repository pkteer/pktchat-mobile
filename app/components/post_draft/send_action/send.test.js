// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';

import Preferences from '@mm-redux/constants/preferences';
import {changeOpacity} from '@utils/theme';

import SendAction from './index';

describe('SendAction', () => {
    const baseProps = {
        testID: 'post_draft.send_action',
        theme: Preferences.THEMES.denim,
        handleSendMessage: jest.fn(),
        disabled: false,
    };

    function getWrapper(props = {}) {
        return shallow(
            <SendAction
                {...baseProps}
                {...props}
            />,
        );
    }

    test('should match snapshot', () => {
        const wrapper = getWrapper();

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should render theme backgroundColor', () => {
        const wrapper = getWrapper();

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.childAt(0)).toHaveStyle('backgroundColor', baseProps.theme.buttonBg);
    });

    test('should change theme backgroundColor to 0.3 opacity', () => {
        const wrapper = getWrapper({disabled: true});

        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.childAt(0)).toHaveStyle('backgroundColor', changeOpacity(baseProps.theme.buttonBg, 0.3));
    });
});
