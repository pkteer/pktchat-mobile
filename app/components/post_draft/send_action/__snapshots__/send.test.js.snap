// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendAction should change theme backgroundColor to 0.3 opacity 1`] = `
<View
  style={
    Object {
      "justifyContent": "flex-end",
      "paddingRight": 8,
    }
  }
  testID="post_draft.send_action.send.button.disabled"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#1c58d9",
          "borderRadius": 4,
          "height": 32,
          "justifyContent": "center",
          "width": 80,
        },
        Object {
          "backgroundColor": "rgba(28,88,217,0.3)",
        },
      ]
    }
  >
    <CompassIcon
      color="rgba(255,255,255,0.5)"
      name="send"
      size={24}
    />
  </View>
</View>
`;

exports[`SendAction should match snapshot 1`] = `
<TouchableWithFeedbackIOS
  onPress={[MockFunction]}
  style={
    Object {
      "justifyContent": "flex-end",
      "paddingRight": 8,
    }
  }
  testID="post_draft.send_action.send.button"
  type="opacity"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1c58d9",
        "borderRadius": 4,
        "height": 32,
        "justifyContent": "center",
        "width": 80,
      }
    }
  >
    <CompassIcon
      color="#ffffff"
      name="send"
      size={24}
    />
  </View>
</TouchableWithFeedbackIOS>
`;

exports[`SendAction should render theme backgroundColor 1`] = `
<TouchableWithFeedbackIOS
  onPress={[MockFunction]}
  style={
    Object {
      "justifyContent": "flex-end",
      "paddingRight": 8,
    }
  }
  testID="post_draft.send_action.send.button"
  type="opacity"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#1c58d9",
        "borderRadius": 4,
        "height": 32,
        "justifyContent": "center",
        "width": 80,
      }
    }
  >
    <CompassIcon
      color="#ffffff"
      name="send"
      size={24}
    />
  </View>
</TouchableWithFeedbackIOS>
`;
